### ISSUE TITLE
stackoverflow with 'lein repl' using leiningen 2.9.7

### ISSUE THREAD

**Issue by torkus** [technomancy/leiningen#2769]
**Initial debugging steps**
Before creating a report, _especially_ around exceptions being thrown when running Leiningen, please check if the error still occurs after:

- [x] Updating to using the latest released version of Leiningen (`lein upgrade`).
- [x] Moving your `~/.lein/profiles.clj` (if present) out of the way. This contains third-party dependencies and plugins that can cause problems inside Leiningen.
  - not present
- [x] Updating any old versions of plugins in your `project.clj`, especially if the problem is with a plugin not working. Old versions of plugins like nREPL and CIDER (as well as others) can cause problems with newer versions of Leiningen.
   - all plugins disabled
- [x] (If you are using Java 9 or newer), updating your dependencies to their most recent versions. Recent JDK's have introduced changes which can break some Clojure libraries.
  - N/A

**Describe the bug**

After upgrading leiningen running `lein repl` causes a stack overflow error.

**To Reproduce**

```
$ lein repl
Exception in thread "Thread-2" java.lang.StackOverflowError
	at clojure.lang.RT.boundedLength(RT.java:1793)
	at clojure.lang.AFn.applyToHelper(AFn.java:148)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.core$apply.invokeStatic(core.clj:669)
	at clojure.core$mapcat.invokeStatic(core.clj:2787)
	at clojure.core$tree_seq$walk__6403$fn__6404.invoke(core.clj:4934)
	at clojure.lang.LazySeq.sval(LazySeq.java:42)
	at clojure.lang.LazySeq.seq(LazySeq.java:51)
	at clojure.lang.RT.seq(RT.java:535)
	at clojure.core$seq__5419.invokeStatic(core.clj:139)
	at clojure.core$concat$fn__5510.invoke(core.clj:727)
	at clojure.lang.LazySeq.sval(LazySeq.java:42)
	at clojure.lang.LazySeq.seq(LazySeq.java:51)
[...]
	at clojure.core$seq__5419.invokeStatic(core.clj:139)
	at clojure.core$concat$cat__5512$fn__5513.invoke(core.clj:736)
	at clojure.lang.LazySeq.sval(LazySeq.java:42)
	at clojure.lang.LazySeq.seq(LazySeq.java:51)
	at clojure.lang.RT.seq(RT.java:535)
^C
```

**Expected behavior**
Obviously I expected a repl.

**Link to sample project**
If relevant, please provide a link to a project or a `project.clj` that others can use to help debug the issue.
https://github.com/ogri-la/strongbox/blob/342737d893a638a355d43cf474e42d65930a2229/project.clj

**Logs**
If applicable, add logs to help explain your problem, including the command that you ran. If the logs are very large, please put them in a [gist](https://gist.github.com/).

**Environment**
- Leiningen Version: [e.g. 2.9.0]. Get this by running `lein version`.
  - Leiningen 2.9.7 on Java 11.0.12 OpenJDK 64-Bit Server VM
- Leiningen installation method:
  - ~Arch AUR using `yay`: `yay -S leiningen`~
  - Arch using: `pacman -S leiningen`
- JDK Version: [e.g. openjdk version "11.0.1"]. Get this by running `java -version`.
  - openjdk version "11.0.12" 2021-07-20
OpenJDK Runtime Environment (build 11.0.12+7)
OpenJDK 64-Bit Server VM (build 11.0.12+7, mixed mode)
- OS: [e.g. Ubuntu 18.04, macOS 10.14].
  - Arch Linux
- Anything else that might be relevant to your problem?
  - followed a general system upgrade

**Additional context**
Add any other context you have about the problem here. Did this work previously on older versions of Leiningen, or older JDK's?

Reverting to `2.9.6` appears to fix the problem: 

    sudo pacman -U /var/cache/pacman/pkg/leiningen-2.9.6-1-any.pkg.tar.zst

**Comment by edporras** [comment]
Ran into the same problem on macOS Catalina, Leiningen installed via homebred, java `1.8.0_301`. Downgrading to `2.9.6` resolved it.

**Comment by justinwoodland** [comment]
I'm seeing the same thing on MacOS Big Sur 11.6, also installed via homebrew, and in a private github actions workflow (ubuntu-latest). Interestingly it doesn't seem to happen on all clojure projects I use lein with, but I haven't been able to track down the differences.

**Comment by itaysabato** [comment]
Just updated to 2.9.7 and got the same issue.

**Comment by itaysabato** [comment]
Going slightly insane trying to rollback to previous version with brew
@edporras do you have the commands handy?

**Comment by itaysabato** [comment]
Thanks, but that did not work...
It seems brew removed the previous installation - I didn't even ask it to install the new lein, it did it by itself when I installed mvn.

**Comment by edporras** [comment]
I recommend setting this so it doesn't auto-clean up:

```sh
HOMEBREW_NO_INSTALL_CLEANUP=1
```

It's been a while since I've tried installing an old version of a Formula and I know it's all changed now but you could try this. I'm typing this off my head so hope it's ok! 

First, add this so it doesn't auto-update brew but once / day:

```
HOMEBREW_AUTO_UPDATE_SECS=86400
```

Don't forget to source the config/restart Terminal to set the vars before continuing. Then, edit the local copy of the Formula :

`/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/leiningen.rb`

and update `url` and `sha256` to match the values from this commit.

https://github.com/Homebrew/homebrew-core/blob/74ee18b20be9826cda6b3e6d6ef5f3907b4ebd1f/Formula/leiningen.rb

Then try `brew install leiningen` again. 

If brew tries to update itself and something craps out, it's probably been longer than a day so, revert the file to the original, do a `brew update`, then make the edits and try the install again.

Don't forget to pin it, then revert the Formula file.

I think it should work but I don't know for sure. Tell me if it works.

**Comment by technomancy** [comment]
Thank you for reporting this. So far we only have one repro case; it would be helpful to get more data to understand the scope of the problem.

The output message specifically advises disabling pedantic mode. When you do that on the one provided repro case, the problem goes away. For the others who are reporting problems, does following the advice given help?

That check was specifically intended for cases where checking pedantic caused crashes: https://github.com/technomancy/leiningen/issues/2763 But it sounds like it's triggering issues in cases where regular pedantic mode previously had no trouble? We need more data to understand how to handle this correctly.

**Comment by itaysabato** [comment]
@edporras Thanks, I ended up copying the whole file, and also removing some cached files.
Seems to have worked in the end.

Seems crazy that brew doesn't manage versions as first-class-citizens at this stage.

**Comment by edporras** [comment]
> Thank you for reporting this. So far we only have one repro case; it would be helpful to get more data to understand the scope of the problem.
> 
> The output message specifically advises disabling pedantic mode. When you do that on the one provided repro case, the problem goes away. For the others who are reporting problems, does following the advice given help?
> 
> That check was specifically intended for cases where checking pedantic caused crashes: #2763 But it sounds like it's triggering issues in cases where regular pedantic mode previously had no trouble? We need more data to understand how to handle this correctly.

Indeed, if I enable the flag in my project, it works with `2.9.7`. I reverted back to `2.9.6` but I'm attaching the error trace in case it helps.

[lein-2.9.7-error.txt](https://github.com/technomancy/leiningen/files/7261722/lein-2.9.7-error.txt)

**Comment by itaysabato** [comment]
I am too scared to upgrade right now, since it was such a pain to rollback last time.
Also, looking at the commit, I fail to see how it can prevent a StackOveflowError...

**Comment by technomancy** [comment]
> I am too scared to upgrade right now, since it was such a pain to rollback last time.

If homebrew is giving you so much trouble then you would probably be better off with a manual install. It's very easy to do and gives you control over exactly which version you use with `lein upgrade` and `lein downgrade`.

But in this case you don't have to replace your current version to test the fix; no rollback required. Just check out the leiningen repository, run `lein install` in the `leiningen-core` directory, and then try the `bin/lein` script from the repo in your codebase.

> looking at the commit, I fail to see how it can prevent a StackOveflowError...

The fix is a combination of that commit plus the previous one: https://github.com/technomancy/leiningen/commit/22eb90cda0636bd7ad155a504b569fa578bd7152

If no one tries the latest version until it's released then it's likely the next release will be have bugs in it.

**Comment by edporras** [comment]
> I believe this is solved in [ec5aeec](https://github.com/technomancy/leiningen/commit/ec5aeece074c3d71f348a1a8127298ce57f8c2da); could anyone who is affected by this please check the latest from git?

Looks to be working fine for me. Tested on several projects and didn't get any errors.

@itaysabato you don't have to uninstall leiningen from homebrew; just unlink it temporarily. Here's what I did:

```sh
git clone git@github.com:technomancy/leiningen.git
cd leiningen/leiningen-core
lein install
lein boostrap
```

(@technomancy I also had to run `lein bootstrap`)

Then I unlinked brew's leiningen:

```sh
brew unlink leiningen
```

I did `which lein` to confirm there's no leiningen in the path. Then I changed into each of my project directories and ran a clean and test to check:

```sh
cd myproject
PATH="$HOME/leiningen/bin:$PATH" lein clean
PATH="$HOME/leiningen/bin:$PATH" lein test
```

Finally, I re-linked brew's Leiningen.

```sh
brew link leiningen
```

**Comment by technomancy** [comment]
Thanks for the confirmation!
